smartapps/smartthings/samsung-smart-tv-connect.src/samsung-smart-tv-connect.groovy

```diff
@@ -0,0 +1,352 @@
+    
+	def nextPage = ""
+    def title = "Linking with your TV"
+	def paragraphText = "Press the 'Yes' button on a message, which is going to be shown on TV screen, to setup a link."
+    
+    if (state.permission)
+    {
+    	nextPage = "tvName"
+		title = "Success! - click 'Next'"
+		paragraphText = "Linking to your hub was a success! Please click 'Next'!"
+    }
+    else if( !state.permission && ((linkRefreshcount % 5) == 0) )
+    {
+    	sendDeveloperReq()
+    }
+    
+    return dynamicPage(name:"tvBtnPush", title:title, nextPage:nextPage, refreshInterval:refreshInterval) {
+    	section("Button Press") {
+        	paragraph """${paragraphText}"""
+        }
+    }
+}
+
+def tvName() {
+    return dynamicPage(name:"tvName", title:"Set Name!", install:true) {
+		section{
+			input "TVname", "text", title:"Give your Samsung Smart TV a name", required:true
+		}
+    }
+}
+
+private discoverTVs() {
+    sendHubCommand(new physicalgraph.device.HubAction("lan discovery urn:samsung.com:device:SmartThingsTvDevice:1", physicalgraph.device.Protocol.LAN))
+}
+
+private sendHubIPReq(devices) {
+    devices.each {
+       	def hubID = it.value.hubID
+       	def vHub = location.hubs.find{it.id == hubID}
+        def body = "{\"request\" : \"set\", \"type\" : \"hubIP\", \"value\" : \"${vHub.localIP + ":" + vHub.localSrvPortTCP}\"}"
+        sendHubCommand(new physicalgraph.device.HubAction([
+        method: "POST",
+        path: "/smartthings/hub",
+        headers: [
+            HOST: ipAddressFromDni(it.value.ip + ":" + it.value.port),
+            /Content-Type/: "application/json"
+        ],
+        body: "${body}\n"],
+        selectedTV))
+    }
+}
+
+private sendDeveloperReq() {
+	def body = "{\"request\":\"set\",\"type\":\"notification\",\"value\":{\"notification id\":\"TVauthentication\",\"buttons\":[\"OK\"],\"title\":\"Inform\",\"icon\":\"\",\"message\":\"\",\"description\":\"Do you want to register your TV?\",\"device id\":\"\"}}"
+    
+	sendHubCommand(new physicalgraph.device.HubAction([
+    	method: "POST",
+		path: "/smartthings/tvfeature",
+		headers: [
+        	HOST: ipAddressFromDni(selectedTV),
+            /Content-Type/: "application/json",
+        ],
+        body: "${body}\n"],
+        selectedTV))
+}
+
+private verifyTV(deviceNetworkId) {
+	sendHubCommand(new physicalgraph.device.HubAction([
+    	method: "POST",
+		path: "/smartthings/deviceinfo",
+		headers:[
+        	HOST: ipAddressFromDni(deviceNetworkId),
+            /Content-Type/: "application/json",
+            /Content-Length/: 0,
+        ]],
+        deviceNetworkId))
+}
+
+private verifyTVs(devices) {
+    devices.each {
+    	verifyTV(it?.value?.ip + ":" + it?.value?.port)
+    }
+}
+
+Map televisionsDiscovered() {
+    def vtelevisions = getVerifiedTVs()
+    
+    def map = [:]
+    vtelevisions.each {
+    	def value = "Samsung Smart TV (${convertHexToIP(it.value.ip)})"
+        def key = it.value.ip + ":" + it.value.port
```

Comment:
The key should be the DNI, which should be based on the MAC address and not the IP:Port.